### PR TITLE
Ref::Util post

### PR DESCRIPTION
**Pull request by xsawyerx**

### CODE REVIEW

File: 2024/incoming/ref-util.pod [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+    use Ref::Util qw< is_hashref >;
+
+    if ( is_hashref($sleigh) ) {
+        load_presents($sleigh);
+    }
+
+Isn't that as refreshing as a winter's first snow? ❄️  The code is cleaner and more readable, but the true magic lies beneath the surface. L<Ref::Util> doesn't just sprinkle a bit of tinsel on your code — it B<supercharges> it!
+
+But wait, there's more under the tree! L<Ref::Util> provides a sleigh full of functions to check all sorts of references:
```
Comment (gonzus) [inline 2024/incoming/ref-util.pod:35]: These two paragraphs overlap a bit.

File: 2024/incoming/ref-util.pod [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+=head2 Performance: The Gift That Keeps on Giving 🎁
+
+Now, you might be thinking, "Sure, the code looks nicer, but is that all?" Oh, dear reader, the true magic lies in the B<massive speed improvements>! L<Ref::Util> doesn't just make your code more readable — it makes it B<faster>.
+
+Let's peek into Santa's workshop and look at some benchmarks:
+
+    ref():                                   5.335e+00 +/- 1.8e-02 (0.3%)
+    ref(), reftype(), !blessed():            1.5545e+01 +/- 3.1e-02 (0.2%)
+    Ref::Util::is_plain_arrayref (CustomOP): 2.7951e+00 +/- 6.2e-03 (0.2%)
+    Data::Util::is_array_ref:                5.9074e+00 +/- 7.5e-03 (0.1%)
```
Comment (gonzus) [inline 2024/incoming/ref-util.pod:111]: This is a bit of a personal peeve of mine, but I **hate** timings printed in scientific notation. I have to carefully check for the exponent first, note that the second entry has a different order of magnitude, and then look at the mantissa while keeping this in mind. Even worse for negative exponents! My brain is much better prepared to look at a list of decimal-point-aligned numbers. Additionally, these numbers should **always** be sorted in such a way that the **best** alternative comes up first!
```
Ref::Util::is_plain_arrayref (CustomOP):  2.7951 +/- 0.0062 (0.2%)
ref():                                    5.3350 +/- 0.0180 (0.3%)
Data::Util::is_array_ref:                 5.9074 +/- 0.0075 (0.1%)
ref(), reftype(), !blessed():            15.5450 +/- 0.0310 (0.2%)
```

File: 2024/incoming/ref-util.pod [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+So, how does L<Ref::Util> achieve this festive feat? It's all thanks to some Christmas magic called B<Custom Ops>, introduced in Perl B<5.14> (released in 2011). Let's unwrap this concept together.
+
+=head3 What Are Op Codes and the Optree?
+
+In Perl, your code gets compiled into a tree of operations called the B<optree>. Each operation (B<op>) represents a fundamental action, like addition, subtraction, or fetching a variable. Think of it as the assembly instructions for Santa's elves.
+
+=head3 The Traditional Method's Optree
+
+When you write:
```
Comment (gonzus) [inline 2024/incoming/ref-util.pod:125]: Maybe rephrase as "This is the optree for a simple one-liner:", or split into two introductions, as you did below: "For a simple one-liner such as XXX, you get the following optree:".

**Review by gonzus — APPROVED** [review]
Some suggestions, nothing too earth shattering.

File: 2024/incoming/ref-util.pod (hunk 2 above)
Comment (xsawyerx) [reply 2024/incoming/ref-util.pod:111]: This is a good point. The version I had was just copy/pasted verbatim from the benchmarking tool. Your version is much more readable.

Comment (gonzus) [reply 2024/incoming/ref-util.pod:111]: Even better:
```
Ref::Util::is_plain_arrayref (CustomOP):  2.7951 +/- 0.0062 (0.2%) -- baseline
ref():                                    5.3350 +/- 0.0180 (0.3%) -- 1.90 times
Data::Util::is_array_ref:                 5.9074 +/- 0.0075 (0.1%) -- 2.11 times
ref(), reftype(), !blessed():            15.5450 +/- 0.0310 (0.2%) -- 5.56 times
```

**Comment by openstrike** [comment]
How about including some mention of [the previous calendar story](https://perladvent.org/2016/2016-12-02.html) on the same module?

**Comment by xsawyerx** [comment]
Oh wow... I didn't even remember writing that. @oalders, do you think it's still worth having this? One was already written in the past, as @openstrike notes. I'm happy to just close the PR.

**Comment by oalders** [comment]
> do you think it's still worth having this?

Of course. 😄

File: 2024/incoming/ref-util.pod (hunk 1 above)
Comment (xsawyerx) [reply 2024/incoming/ref-util.pod]: One is more about the speed, another is about the functions. I merged them in a cleaner way.

File: 2024/incoming/ref-util.pod (hunk 2 above)
Comment (xsawyerx) [reply 2024/incoming/ref-util.pod]: Added.

File: 2024/incoming/ref-util.pod (hunk 3 above)
Comment (xsawyerx) [reply 2024/incoming/ref-util.pod]: Done.

**Comment by xsawyerx** [comment]
Updated based on feedback. Included note to previous post.

**Review by oalders — APPROVED** [review]
⭐⭐⭐⭐⭐

Thanks, @xsawyerx! Very enjoyable to read.
